### PR TITLE
Add in updates to ERC7535 + Oracle

### PR DESCRIPTION
**Pull request by Tranquil-Flow**

### CODE REVIEW

File: src/vault/ERC7535MultiRewardVault.sol [hunk 1]
```diff
@@ -204,21 +208,22 @@ contract ERC7535MultiRewardVault is ERC7535Upgradeable, OwnableUpgradeable, Reen
...
      */
     function totalAssets() public view override returns (uint256) {
-        // Calculate value of assets (native ETH)
-        uint256 assetAmount = address(this).balance;
-        address asset = address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE); // Native ETH
-        TokenInfo memory ethInfo = assetInfo[IERC20MetadataUpgradeable(asset)];
-        uint256 totalValue = assetAmount * oracle.getPrice(asset, ethInfo.priceFeed);
+        // Calculate value of native ETH
+        uint256 ethBalance = _getETHBalance();
+        uint256 ethPrice = oracle.getPrice(address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE), address(0));
```
Comment (asanson1404) [inline src/vault/ERC7535MultiRewardVault.sol:217]: do not use address(0) for the ETH / USDC priceFeed

File: src/vault/ERC7535MultiRewardVault.sol [hunk 2]
```diff
@@ -229,26 +234,34 @@ contract ERC7535MultiRewardVault is ERC7535Upgradeable, OwnableUpgradeable, Reen
...
     function _convertToShares(uint256 assets, MathUpgradeable.Rounding rounding) internal view override returns (uint256 shares) {
-        uint256 supply = totalSupply();
-        return (supply == 0)
-            ? assets
-            : assets.mulDiv(supply, totalAssets(), rounding);
+        uint256 supply = totalSupply() + 10 ** _decimalsOffset(); // Supply includes virtual reserves
+        if (totalSupply() == 0) {
+            return assets; // On first deposit, totalSupply is 0, so return assets (amount of ETH deposited) as shares
+        } else {
+            uint256 ethPrice = oracle.getPrice(address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE), address(0));
```
Comment (asanson1404) [inline src/vault/ERC7535MultiRewardVault.sol:246]: Same here... Why do you use address(0)?

File: src/vault/ERC7535MultiRewardVault.sol [hunk 3]
```diff
@@ -229,26 +234,34 @@ contract ERC7535MultiRewardVault is ERC7535Upgradeable, OwnableUpgradeable, Reen
...
-        uint256 supply = totalSupply();
-        return (supply == 0)
-            ? shares
-            : shares.mulDiv(totalAssets(), supply, rounding);
+        uint256 supply = totalSupply() + 10 ** _decimalsOffset(); // Supply includes virtual reserves
+        if (totalSupply() == 0) {
+            return shares; // If there are no shares, return the number of shares as assets. TODO: Remove unnecessary code?
+        } else {
+            uint256 assetsUsdValue = shares.mulDiv(totalAssets(), supply, rounding);
+            uint256 ethPrice = oracle.getPrice(address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE), address(0));
```
Comment (asanson1404) [inline src/vault/ERC7535MultiRewardVault.sol:262]: Same here, change the priceFeed address

**Review by asanson1404 — CHANGES_REQUESTED** [review]

**Review by asanson1404 — CHANGES_REQUESTED** [review]
Just saw your commit, the code is clean well done Evi :)

I just don't understand why the proxy / priceFeed address on the function `oracle.getPrice()` is still `address(0)`?
Also, don't you think it's better to change the prototype of  `oracle.getPrice` to only have an argument, the token address. The priceFeed / proxy of each token could be store in the OracleImplementation contract.

Also, I thought a bit about that priceFeed vulnerability... How could we be sure that the vault creator will use the correct price feed for the correct tokens and won't use a wrong price feed to boost his rewards or break the vault.
Perhaps, oracle-agnostic means to choose OracleImplementation among a set which has already default and immutable price feed?

I think that would be nice to check the Morpho contracts to see how do they handle that Oracle-agnostic feature

Just saw the interface of Morpho here (no price feed specified):
https://github.com/morpho-org/morpho-blue/blob/fcb190b0e0e1355defe56b19781ac573986c3f74/src/interfaces/IOracle.sol

https://docs.morpho.org/morpho/contracts/oracles/
https://docs.morpho.org/morpho/contracts/oracles/chainlink-oracle-adapter
https://docs.morpho.org/morpho/concepts/oracle-agnostic-pricing
https://docs.morpho.org/morpho/concepts/risk-documentation#oracle-risk

**Comment by Tranquil-Flow** [comment]
`oracle.getPrice(address token, address priceFeed)` still uses address(0) for `priceFeed` as no value is needed in this call. In the OracleImplementation, if token = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE , then it uses the hardcoded ETH-USD price feed. `priceFeed` does nothing in this call.

Removing the 2nd parameter on the `getPrice` function on IOracle is not a bad idea if we are going with this approach of storing the price feeds for specific tokens on the oracle contracts. It does mean that every new reward token to the vault must have it's price feed stored on the oracle implementation the contract is using - either during the function call to add a reward token or it being an immutable stored variable. In return, it means that oracles are less flexible. They would need a stored price feed to correspond to an address - where as if you pass in a price feed an oracle implementation should work for any price feed that oracle supports.

Morpho has price feed's specified as immutable variables for their `MorphoChainlinkOracleV2` contract which implements their `IOracle`. Basically oracle-agnostic just means that any oracle source/contract can be used as long as calling `price` returns the price given a defined collateral and loan token.

**Pull request merged**